src/quacc/atoms/skzcam.py

```diff
@@ -19,6 +19,420 @@
+    slab_indices = [
+        i
+        for i in range(len(quantum_cluster))
+        if quantum_cluster.get_array("atom_type")[i] != "adsorbate"
+    ]
+
+    # Get the charge of the quantum cluster
+    charge = int(sum(quantum_cluster.get_array("oxi_states")))
+
+    # Create the coords strings for the adsorbate-slab complex
+    ad_slab_coords = f"""%coords
+CTyp xyz
+Mult {multiplicity['ad_slab']}
+Units angs
+Charge {charge}
+coords
+"""
+    for i in range(len(quantum_cluster)):
+        position = quantum_cluster[i].position
+        ad_slab_coords += f"{quantum_cluster.get_chemical_symbols()[i].ljust(3)} {' '*16} {position[0]:-16.11f} {position[1]:-16.11f} {position[2]:-16.11f}\n"
+
+    # Create the coords section for the ECP region
+    ecp_region_coords_section = ""
+    for i in range(len(ecp_region)):
+        position = ecp_region[i].position
+        pc_charge = ecp_region.get_array("oxi_states")[i]
+        ecp_region_coords_section += f"{(ecp_region.get_chemical_symbols()[i] + '>').ljust(3)} {position[0]:-16.11f} {pc_charge:-16.11f} {position[1]:-16.11f} {position[2]:-16.11f}\n"
+        if (
+            ecp_info is not None
+            and ecp_region[i].symbol in ecp_info
+            and ecp_info[ecp_region[i].symbol] is not None
+        ):
+            atom_ecp_info = format_ecp_info(ecp_info[ecp_region[i].symbol])
+            ecp_region_coords_section += f"{atom_ecp_info}"
+
+    # Add the ECP region coords section to the ads_slab_coords string
+    ad_slab_coords += ecp_region_coords_section
+    ad_slab_coords += "end\nend\n"
+
+    # Create the coords string for the adsorbate
+    ad_coords = f"""%coords
+CTyp xyz
+Mult {multiplicity['ad']}
+Units angs
+Charge 0
+coords
+"""
+    for i in range(len(quantum_cluster)):
+        if i in adsorbate_indices:
+            position = quantum_cluster[i].position
+            ad_coords += f"{quantum_cluster.get_chemical_symbols()[i].ljust(3)} {' '*16} {position[0]:-16.11f} {position[1]:-16.11f} {position[2]:-16.11f}\n"
+        elif include_cp:
+            position = quantum_cluster[i].position
+            ad_coords += f"{(quantum_cluster.get_chemical_symbols()[i]+':').ljust(3)} {' '*16} {position[0]:-16.11f} {position[1]:-16.11f} {position[2]:-16.11f}\n"
+    ad_coords += "end\nend\n"
+
+    # Create the coords string for the slab
+    slab_coords = f"""%coords
+CTyp xyz
+Mult {multiplicity['slab']}
+Units angs
+Charge {charge}
+coords
+"""
+    for i in range(len(quantum_cluster)):
+        if i in slab_indices:
+            position = quantum_cluster[i].position
+            slab_coords += f"{quantum_cluster.get_chemical_symbols()[i].ljust(3)} {' '*16} {position[0]:-16.11f} {position[1]:-16.11f} {position[2]:-16.11f}\n"
+        elif include_cp:
+            position = quantum_cluster[i].position
+            slab_coords += f"{(quantum_cluster.get_chemical_symbols()[i] + ':').ljust(3)} {' '*16} {position[0]:-16.11f} {position[1]:-16.11f} {position[2]:-16.11f}\n"
```

Comment:
It seems like nearly the same loop pattern is being done 3 times. To reduce code duplication, it may be worth factoring out the common bits to a dedicated function. If it's an internal function with no user-facing purpose, you can prepend the function name with `_` in the front so it's not rendered in the docs.

In any case, let's try to reduce duplication throughout this function if possible.

Reply:
I've now added a dedicated function: `create_atom_coord_string` to write out the coordinate string when given an Atom object.